nexus/saga-recovery/src/recovery.rs

```diff
@@ -0,0 +1,705 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Guts of the saga recovery bookkeeping
+
+use super::status::RecoveryFailure;
+use super::status::RecoverySuccess;
+use chrono::{DateTime, Utc};
+use omicron_common::api::external::Error;
+use slog::{debug, error, info, warn};
+use slog_error_chain::InlineErrorChain;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use steno::SagaId;
+use tokio::sync::mpsc;
+
+/// Describes state related to saga recovery that needs to be maintained across
+/// multiple passes
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct RestState {
+    /// set of sagas that we believe may be running
+    ///
+    /// See the big block comment in the saga recovery background task for more
+    /// on how this works and why.
+    sagas_started: BTreeMap<SagaId, SagaStartInfo>,
+    remove_next: BTreeSet<SagaId>,
+}
+
+/// Describes how we learned that a particular saga might be running
+///
+/// This is only intended for debugging.
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[allow(dead_code)]
+struct SagaStartInfo {
+    time_observed: DateTime<Utc>,
+    source: SagaStartSource,
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum SagaStartSource {
+    StartChannel,
+    Recovered,
+}
+
+impl RestState {
+    /// Returns initial at-rest state related to saga recovery
+    pub fn new() -> RestState {
+        RestState {
+            sagas_started: BTreeMap::new(),
+            remove_next: BTreeSet::new(),
+        }
+    }
+
+    /// Read messages from the channel (signaling sagas that have started
+    /// running) and update the set of sagas that we believe may be running.
+    ///
+    /// See the big block comment in the saga recovery background task for more
+    /// on how this works and why.
+    pub fn update_started_sagas(
+        &mut self,
+        log: &slog::Logger,
+        sagas_started_rx: &mut mpsc::UnboundedReceiver<SagaId>,
+    ) -> usize {
+        let (new_sagas, disconnected) = read_all_from_channel(sagas_started_rx);
+        if disconnected {
+            warn!(
+                log,
+                "sagas_started_rx disconnected (is Nexus shutting down?)"
+            );
+        }
+
+        let rv = new_sagas.len();
+        let time_observed = Utc::now();
+        for saga_id in new_sagas {
```

Comment:
What's the motivation for reading all the messages into a `Vec` and _then_ looping over them and putting them in the `BTreeMap`? I get that we want to be able to write unit tests for `read_all_from_channel`, but it occurs to me that we could still have that be factored out into a function and instead have it take an `&mut BTreeMap`. That way, we should still be able to write tests for it, but avoid having to loop twice.

Not a big deal, though.

Reply:
There's no significant motivation here but I think it makes this cleaner.  We could have `read_all_from_channel()` accept a `&mut BTreeSet` (I assume set here and not map because these are just ids?), but then a reader might reasonably wonder if it _must_ be empty, whether the elements in it might be removed or mutated, what happens if a key is already present, what happens if two items on the channel compare equal, etc.  At that point it'd be clearer to just inline the code into the one place it's used, but I slightly prefer what's here if that's ok.